### ISSUE TITLE
Focus management for Multi-Screen Sims

### ISSUE THREAD

**Issue by terracoda** [phetsims/ratio-and-proportion#282]
**On Desktop**
- When I load a multi-screen sim focus is placed on the H1 of the home page with no visible pink focus highlight. (This has been good for interview participants) 
- When I choose a sim screen from the Home Screen, focus starts on the H1 of the new screen, again with no visible pink focus high light. (This has been good for interview participants)
- When I choose a sim screen from the bottom navbar, focus remains on the selected sim screen button. (This has been surprising to participants.)
- When I choose the Home Screen from the bottom navbar, focus is placed on the most recently active screen. (I have no data on this.)

I have noticed in interviews with blind users that they find it surprising (and are often temporarily disorientated) when they load the second sim screen from the bottom navbar and focus has stayed on the active sim screen button (i.e., not moved to the top of the new screen). 

In one Ratio and Portion interview, one blind participant explicitly said that focus **should** go to the top of the screen. 

This is the first fully described multi-screen sim that will be published. Coulomb's Law was published with alternative input a few years back, and Waves Intro will be published with alternative input in the near future.

**Focus behaviour is different on iOS with VO enabled**
- When I load the sim, focus correctly goes on the H1.
- When I choose the Discover Screen from home screen, focus seems to always land on the Right Hand slider.
- When I choose the Create screen from home screen, focus seems to always land on No Tick Marks radio button.
- When choosing a screen from the navbar, focus stays on the active screen button.

I have only done one interview with RaP on a mobile device, and we had other issues, so not a lot of good data and screen navigation.

**Design Question**
For sims with Interactive Description, can we easily place focus on the H1 of the active screen in all cases and on all platforms? I'm thinking even when returning to the Home Screen, we could place focus on the H1. I don't think it would be problematic, and it would introduce a lot of consistency for non-visual access.

**Comment by terracoda** [comment]
@zepumph and @jessegreenberg could either of you comment on how straightforward it would be from a code and implementation perspective to make focus always go to the H1 when a screen change happens?

More generally, @emily-phet do you think we need to have a wider (brief) discussion on focus management of multi-screen sims?

**Comment by zepumph** [comment]
>When choosing a screen from the navbar, focus stays on the active screen button.

This feels like some we should change, and that would be pretty straight forward to do.



>Focus behaviour is different on iOS with VO enabled
>. . . 
>When I choose the Discover Screen from home screen, focus seems to always land on the Right Hand slider.
>When I choose the Create screen from home screen, focus seems to always land on No Tick Marks radio button.

This is peculiar and bad. Does this behavior change depending on what rotary function VO is on, like "form controls" vs "lines"

**Comment by terracoda** [comment]
Regarding @zepumph's question
> This is peculiar and bad. Does this behavior change depending on what rotary function VO is on, like "form controls" vs "lines"

I think I am using lines as I have not turned on the rotar.

**Comment by emily-phet** [comment]
@terracoda I trust your judgement on where you would like focus to go. If this is feasible, and pretty straightforward from a development perspective, let's do it. If it's more challenging or will require some investigation, let's incorporate it into our planning work in January  - so someone put it on your list (see email coming later today) of things you'd like to have happen in 2021.

**Comment by jessegreenberg** [comment]
I have to admit, I think this will be a little confusing for sighted users. From their perspective the focus will simply disappear when they make a selection. If they just want to click through screen buttons they will have to tab through each screen to do so. I think this goes against the "don't move focus for the user" approach we have generally followed and makes assumptions about what the user wants to do.

But if we do want to do this, could we just blur the button in the nav bar rather than focus the h1 of the next screen? That would match behavior for single screen sims and screens started from home screen when using a screen reader.

**Comment by terracoda** [comment]
@jessegreenberg, I am always a little confused about what "blur" means.

If I am using a screen reader and I choose screen 2 from the navbar, and then hit the Tab key, where would I go?
If I am using a screen reader and I choose screen 2 from the navbar, and then hit the the down arrow key, where would I go?

**Comment by jessegreenberg** [comment]
By "blur" I just mean removing focus without putting it anywhere else.  Focus and the virtual cursor should then start from the top of the page.

> If I am using a screen reader and I choose screen 2 from the navbar, and then hit the Tab key, where would I go?

It would go to the first focusable thing in the screen.

> If I am using a screen reader and I choose screen 2 from the navbar, and then hit the the down arrow key, where would I go?

It should start reading from the top of the page, presumably the "H1" for the screen. 

Behavior should be just like landing on the screen from the Home Screen.

**Comment by terracoda** [comment]
Ooops, I responded yesterday, but must have forgotten to press the comment button.

@jessegreenberg, what you describe above seems to be what I think users would expect. In interviews with blind learners, I have rarely witnessed trouble, surprise, uncertainty when a sim loads, or more recently with RaP when a learner navigates to a screen from the Home Page.

However, when navigating to a second screen from the navbar, I have witnessed trouble in one form or another.

Can we try this "blur" approach? I think this would enable blind learners to start at the top on each screen load without disrupting visual learners. I think visual keyboard learners would also likely want to start on the first important interactive object. Pressing the tab key after the new screen loads would get them easily on that first interactive object.

**Comment by terracoda** [comment]
I'll add, that in the past, I have thought that the sim screens were somewhat like tab panels and the sim screen buttons somewhat like "tabs". I think that this blur approach might mimic that type of navigation experience while maintaining our own language of "Home Screen", {{Sim}} Screen button", and "{{Sim}} Screen" rather than introducing tabpanel and tab semantics. 

I think our "screen" semantics are clearly and effectively communicating the multi-screen sim experience. My only concern is this focus issue when changing screen from the Sim Screens navbar. Blurring focus once a new screen is chosen sounds like what we need.

**Comment by terracoda** [comment]
Tagging @jessegreenberg in case you and @zepumph will work on this together.

**Comment by zepumph** [comment]
@jessegreenberg, bluring did not achieve what we wanted. When I blurred and tested with NVDA, the virtual cursor still stayed right on the screen button in the nav bar.

```diff
Index: js/NavigationBarScreenButton.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/NavigationBarScreenButton.js b/js/NavigationBarScreenButton.js
--- a/js/NavigationBarScreenButton.js	(revision e0b018e279d3011821f909d1dc0f362f7a573454)
+++ b/js/NavigationBarScreenButton.js	(date 1608133192445)
@@ -121,6 +121,8 @@
     this.buttonModel = new PushButtonModel( {
       listener: () => {
         screenProperty.value = screen;
+        this.blur();
+        // document.body.focus(); // perhaps add this in addition?
       },
       tandem: options.tandem,
 

```

I am unsure about the strategies that are usually used to try to "refresh" a page to focus back to the top of the screen. I also tried a blanket focus of the document body and that didn't work either. Any thoughts here?

**Comment by terracoda** [comment]
@zepumph, what's going on with focus when a learner hits a screen button on the home page?

**Comment by jessegreenberg** [comment]
> By "blur" I just mean removing focus without putting it anywhere else. Focus and the virtual cursor should then start from the top of the page.

It looks like I was totally incorrect about this. Not only the virtual cursor but the DOM focus will also move to the next thing in order on tab press even after an element is blurred. If you click in empty space in a webpage, focus will start over and go back to the top of the page. I thought this was because of the blur but I guess it is because of the user click.

>  what's going on with focus when a learner hits a screen button on the home page?

I guess it is just because of order of elements in the DOM. When user clicks a screen button on the home page, the elements on the home screen become hidden, and the button gets blurred. The active screen becomes un-hidden, and focus enters it from the top because it comes after the home screen naturally in DOM order.

The reason I was hoping `blur()` would work was because we currently don't have a way to focus a Node's label sibling element (the h1). We could change the scene graph in ScreenView to make the H1 its own Node.  Then hopefully we can focus this Node, even though it doesn't have defined graphical bounds (I think I recall this being a limitation before, but maybe there is a way). @zepumph what do you thin, do you see another way forward?

**Comment by zepumph** [comment]
I think you are on to something here, and that this is worth investigation:

>We could change the scene graph in ScreenView to make the H1 its own Node. Then hopefully we can focus this Node, even though it doesn't have defined graphical bounds (I think I recall this being a limitation before, but maybe there is a way).

My bad idea was to potentially rearrange the articles of the PDOM dynamically to achieve the right next-focusable item, but that feels pretty heavyweight. Maybe we can play around with something focusable that exists as the first thing on each screen. @jessegreenberg let's brainstorm tomorrow during out meeting.

**Comment by zepumph** [comment]
@jessegreenberg and I solved this today. We were able to generally focus the screen when a screen becomes visible. This was just a single listener in Screen.js, as well as hiding the focus highlight for a screen view to protect this feature from visual side-effects. We found one small bug in AnimatedPanZoomListener from this because of the initial focus on startup.

@jessegreenberg found one side-effect worth further discussion, where if you select a nav bar screen button while zoomed in, the focus change will cause a pan to the center of the screen. He created phetsims/joist#682 to solve that in a more isolated way.

@terracoda we also found that with a screen reader, selecting a new screen will cause the sim to automatically read the screen summary. Is this acceptable?

Please give it a try and let us know what you think!

**Comment by terracoda** [comment]
Awesome. Will review in the morning!

**Comment by terracoda** [comment]
@zepumph and @jessegreenberg, I am really liking the behaviour from a screen reader behaviour. Here's why:

1. I load a screen from the home page...
2. The page starts reading from the top.
3. I can verify it is the top by pressing Ctrl+Opt+left arrow - that takes me to the previous line, an H1.
4. I can then continue reading from automatically from top, or at my own pace with the cursor key (Ctrl+Opt+ right arrow on VO)
5. I can press tab and land on the first (and usually most important) interactive object.

I think this a great flow.
I don't think auto-reading from the top is an issue. SR users can easily stop it and continue at their own pace.

For visual keyboard users, pressing the Tab key after the screen loads puts a learner's focus on the first (and usually most important) interactive object. I think this is good design.

We don't have similar behaviour for the Home screen. Focus lands on the most recently active screen button. I think this may actually be a useful thing.

**My thoughts:**
- I don't actually have user data on the home, but from my perspective, the home page has minimal content that generally would only need to be read once. 
- I don't think I have ever witnessed a single non-visual user ever returning to the home screen, though I do remember saying to one interviewee that it was not necessary to return to the home screen in order to go to the next screen.
- since the main reason to go to the Home screen would be to find another screen, I think it is fine that focus lands on the screen button. Future research will determine if I am wrong about this.

Nice work, @zepumph and @jessegreenberg! Closing this issue.

**Comment by zepumph** [comment]
I'm not sure we worked on the iOS behavior in this issue at all. I just found that pressing a nav bar screen button didn't focus the top of the sim. I'm not sure this needs to block publication though.

**Comment by zepumph** [comment]
What do you think @terracoda?

**Comment by terracoda** [comment]
Oh darn, I just verified again for keyboard and screen reader, and of course it works.

I also noticed that this was not working on iOS with VO enabled.
I think it should apply there, too.

Focus should go on the top (H1).

Currently, focus seems a little erratic, so setting in on the H1 when the screen loads, might help.

**Comment by terracoda** [comment]
Actually, wait a second. This might not be right for iOS without VO enabled.
And you can't tell if a screen reader is enabled.
Reassigning to myself with I re-think this briefly.

**Comment by terracoda** [comment]
As I said before, focus management is working great on my laptop with VO.

Now for my phone...
Ok, no focus highlights show up as I use the sim with regular touch gestures with VO off. Focus doesn't really exist in this exploration.

When I enable VO, and use VO's custom gestures (swipes and double taps, etc) I see both our interactive highlight and VO's black highlight. Sometimes they are not aligned, but I think they are doing what they are supposed to be doing.

When I go to a screen from the Home screen, the new screen seems start at the top pretty regularly. This is good.

When I got to a new screen from the navbar, focus stays on the screen button in the navbar.
I can do a two-finger swipe up and VO starts reading from the top of the screen.


While I have been told by screen reader users using the keyboard that the screen should always start at the top, I don't have any data for VO users on this.

My feeling is that moving focus to the top when changing screens from the navbar would not be disruptive, but I am not 100% sure.

Let's hold off on doing anything special for iOS right now. I think I need more data on this, and it doesn't seem broken to me.

**Comment by zepumph** [comment]
Noting that https://github.com/phetsims/ratio-and-proportion/issues/321 was reopened because NVDA was starting from the second paragraph of the screen summary.
